osu.Game/Rulesets/Mods/ModMetronome.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Audio.Track;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Beatmaps.ControlPoints;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Containers;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.UI;
+using osu.Game.Skinning;
+using osu.Game.Configuration;
+using osu.Framework.Bindables;
+using osu.Game.Audio;
+
+namespace osu.Game.Rulesets.Mods
+{
+    public abstract class ModMetronome : Mod
+    {
+        public override string Name => "Metronome";
+        public override string Acronym => "MT";
+        public override IconUsage? Icon => OsuIcon.ModAuto; // temporary icon
+        public override ModType Type => ModType.DifficultyReduction;
+        public override string Description => "Add a metronome";
+        public override bool Ranked => false;
+        public override double ScoreMultiplier => 0.0;
+
+        public enum TickFrequency
+        {
+            One = 1,
+            Two = 2,
+            Four = 4
+        }
+
+        [SettingSource("Tick frequency", "Number of metronome ticks per beat")]
+        public Bindable<TickFrequency> Frequency { get; } = new Bindable<TickFrequency>(TickFrequency.One);
+    }
+
+    public abstract class ModMetronome<TObject> : ModMetronome, IApplicableToDrawableRuleset<TObject>
+        where TObject : HitObject
+    {
+        public void ApplyToDrawableRuleset(DrawableRuleset<TObject> drawableRuleset)
+        {
+            drawableRuleset.Overlays.Add(new MetronomeBeatContainer(Frequency.GetBoundCopy()));
```

Comment:
Generally for safety you would call `GetBoundCopy()` in the receiver, not pass it down like this. That ensures a lack of memory leaks due to possibly keeping references alive through the callback closure.